### PR TITLE
Propagate linkmaps into `xcarchives`

### PR DESCRIPTION
**Pull request by maxwellE**
These files should be propagated into xcarchives if enabled

Related #2279

### CODE REVIEW

**Review by luispadron — APPROVED** [review]
Sweet!

**Review by mattrobmattrob — APPROVED** [review]
Can you add anything to `test/starlark_tests/xcarchive_tests.bzl` to test this new functionality?

**Comment by maxwellE** [comment]
> Can you add anything to `test/starlark_tests/xcarchive_tests.bzl` to test this new functionality?

Will do

**Comment by maxwellE** [comment]
> Can you add anything to `test/starlark_tests/xcarchive_tests.bzl` to test this new functionality?

How can I force a target to build linkmaps outside of passing a flag? I think that is only controlled via a flag right now, I could be mistaken

**Comment by luispadron** [comment]
> How can I force a target to build linkmaps outside of passing a flag? I think that is only controlled via a flag right now, I could be mistaken

You could add a transition to force that but im not sure how that works with the tests to be fair.

**Comment by maxwellE** [comment]
> > How can I force a target to build linkmaps outside of passing a flag? I think that is only controlled via a flag right now, I could be mistaken
> 
> You could add a transition to force that but im not sure how that works with the tests to be fair.

Do you have an example on hand of this? It appears the dsyms are also not tested

**Comment by luispadron** [comment]
Heres how I did it for the `docc_archive` rule and the `swift.emit_symbol_graph` flag: https://github.com/bazelbuild/rules_apple/blob/ba7c7abefbfa3e868a3f7bfa93a98613e2dd70c5/apple/internal/docc.bzl#L155-L168

We use a transition to always set that feature and then an aspect which collects some providers that the rule can then use: https://github.com/bazelbuild/rules_apple/blob/ba7c7abefbfa3e868a3f7bfa93a98613e2dd70c5/apple/internal/aspects/docc_archive_aspect.bzl

I don't think this is required though if we want to keep it as requiring a flag for now, but it might be nice to transition on these always if thats what Xcode does.

**Comment by luispadron** [comment]
I should've probably added the dsym transition when i originally implemented this, today we just set that flag when building for release which works out.

**Comment by maxwellE** [comment]
@luispadron So that looks good but the flag is not a feature in rules apple or rules swift, its in [bazel core ](https://bazel.build/reference/command-line-reference#flag--objc_generate_linkmap) . How do I set that?

**Comment by mattrobmattrob** [comment]
My guess is that you'd need to update `apple_verification_test` to include an argument that allows you to override `--objc_generate_linkmap`. Here's an example that's similar but for `--apple_generate_dsym`.

The rule attribute for `apple_verification_test`:

https://github.com/bazelbuild/rules_apple/blob/ba7c7abefbfa3e868a3f7bfa93a98613e2dd70c5/test/starlark_tests/rules/apple_verification_test.bzl#L271-L276

Which is then used in a transition in the test implementation to control the `--apple_generate_dsym` value.

https://github.com/bazelbuild/rules_apple/blob/ba7c7abefbfa3e868a3f7bfa93a98613e2dd70c5/test/starlark_tests/rules/apple_verification_test.bzl#L67

And then an example how it's used:

https://github.com/bazelbuild/rules_apple/blob/ba7c7abefbfa3e868a3f7bfa93a98613e2dd70c5/test/starlark_tests/ios_framework_tests.bzl#L440-L444

**Comment by luispadron** [comment]
Nice @mattrobmattrob, must've missed that attr when I added this test originally since it's missing a test for dsym. 

@maxwellE if you just want to add another attr to exercise the feature in the test that's fine and I'll follow up to update the rule so we transition these flags automatically (if that's what Xcode does by default)

**Comment by maxwellE** [comment]
@mattrobmattrob Thank you for that, was great to see how to enable that in the future.

Added a test case to cover the logic

File: tools/xcarchivetool/make_xcarchive.py [hunk 1]
```diff
@@ -135,6 +145,14 @@ def _main(
       os.path.join(dsyms_path, os.path.basename(dsym)),
       dirs_exist_ok=True)
 
+  # Copy the linkmap files into the .xcarchive.
+  linkmaps_path = os.path.join(output_path, "Linkmaps")
+  os.makedirs(linkmaps_path, exist_ok=True)
```
Comment (mattrobmattrob) [inline tools/xcarchivetool/make_xcarchive.py:150]: Can you avoid this by following the dSYM `dirs_exist_ok` param usage in the copy call?

Comment (maxwellE) [reply tools/xcarchivetool/make_xcarchive.py:150]: Will do

Comment (tinder-maxwellelliott) [reply tools/xcarchivetool/make_xcarchive.py:150]: Linkmaps are single files though, copyfile does not take that parameter

Comment (maxwellE) [reply tools/xcarchivetool/make_xcarchive.py:150]: Linkmaps are single files though, that argument does not work for copyfile

Comment (mattrobmattrob) [reply tools/xcarchivetool/make_xcarchive.py]: 🤦 Makes sense. Can you avoid creating the directory unless linkmap files are present?

**Pull request merged**